feat(docs): CDC blog post

amyshwang opened the pull request.

amunra commented on blog/2022-12-28-change-data-capture-with-questdb-and-debezium.mdx:
```diff
@@ -0,0 +1,332 @@
+
+The important things to note here are:
+
+- symbols : Stock symbols are translated to
```
Backtick quoting of `symbols`, removal of space before `:`.

amyshwang replied:
```suggestion
- `symbols`: Stock symbols are translated to
```

amunra commented on blog/2022-12-28-change-data-capture-with-questdb-and-debezium.mdx:
```diff
@@ -0,0 +1,332 @@
+  [QuestDB symbol type](/docs/concept/symbol/), used for
+  string values with low cardinality (e.g., enums).
+
+- timestamp.field.name : Since QuestDB has great support for timestamp and
```
Ditto.

amyshwang replied:
```suggestion
- `timestamp.field.name`: Since QuestDB has great support for timestamp and
```

amunra commented on blog/2022-12-28-change-data-capture-with-questdb-and-debezium.mdx:
```diff
@@ -0,0 +1,332 @@
+- timestamp.field.name : Since QuestDB has great support for timestamp and
+  partitioning based on that, we can specify the designated timestamp column.
+
+- transforms : unwrap field uses io.debezium.transforms.ExtractNewRecordState
```
Ditto.

amyshwang replied:
```suggestion
- `transforms`: unwrap field uses `io.debezium.transforms.ExtractNewRecordState`
```

amunra replied:
Backtick-quote `payload.after`

amunra commented on blog/2022-12-28-change-data-capture-with-questdb-and-debezium.mdx:
````diff
@@ -0,0 +1,332 @@
+  other words, this is a filter to basically take the payload.after portion of
+  the Debezium data on the Kafka topics:
+
+```
+  { "schema": { "comment": "this contains Debezium message schema, it's not very
+  relevant for this sample" }, "payload": { "before": null, "after": { "id": 8,
+  "symbol": "NFLX", "price": 1544.3357414199545, "last_update": 1666172978269856
+  }, "source": { "version": "1.9.6.Final", "connector": "postgresql", "name":
+  "dbserver1", "ts_ms": 1666172978272, "snapshot": "false", "db": "postgres",
+  "sequence": "[\"87397208\",\"87397208\"]", "schema": "public", "table":
+  "stock", "txId": 402087, "lsn": 87397208, "xmin": null }, "op": "u", "ts_ms":
+  1666172978637, "transaction": null } }
+```
````
What is this random piece of JSON?

There's no text above it telling me what it is. I'm pretty confused.

amunra replied:
Run
```
jq . file.json
```

to pretty-print the JSON, add a sentence on top explaining what this is (still confusing after just adding a title, IMO).

Is this what is being sent over the network?

amunra commented on blog/2022-12-28-change-data-capture-with-questdb-and-debezium.mdx:
````diff
@@ -0,0 +1,332 @@
+  1666172978637, "transaction": null } }
+```
+
+### QuestDB andGrafana
````
Typo: Add space.
Should read
```
### QuestDB and Grafana
```

amyshwang replied:
```suggestion
### QuestDB and Grafana
```

amunra commented on blog/2022-12-28-change-data-capture-with-questdb-and-debezium.mdx:
```diff
@@ -0,0 +1,332 @@
+### QuestDB andGrafana
+
+Once the data is written to QuestDB tables, we can work with the time-series
+data easier. Since QuestDB is compatible with PostgreSQL wire protocol, we can
```
Grammar: compatible with _the_ PostgreSQL wire protocol.

amyshwang replied:
```suggestion
data easier. Since QuestDB is compatible with the PostgreSQL wire protocol, we can
```

amunra commented on blog/2022-12-28-change-data-capture-with-questdb-and-debezium.mdx:
```diff
@@ -0,0 +1,332 @@
+
+Once the data is written to QuestDB tables, we can work with the time-series
+data easier. Since QuestDB is compatible with PostgreSQL wire protocol, we can
+use the PostgreSQL datasource on Grafana to visualize the data. The preconfigure
```
Typo: preconfigured  (last `d` is missing).

amyshwang replied:
```suggestion
use the PostgreSQL datasource on Grafana to visualize the data. The preconfigured
```

amunra commented on blog/2022-12-28-change-data-capture-with-questdb-and-debezium.mdx:
```diff
@@ -0,0 +1,332 @@
+
+## Next steps
+
+This sample project is a nice reference architecture to stream data from a
```
I think the word "nice" should be substituted with one which is a little more meaningful.

amunra replied:
Either that, or just get rid of it.

amunra replied:
Maybe:

```
This sample project provides a foundational reference architecture ...
```

amyshwang replied:
```suggestion
This sample project is a foundational reference architecture to stream data from a
```

amunra commented on blog/2022-12-28-change-data-capture-with-questdb-and-debezium.mdx:
```diff
@@ -0,0 +1,332 @@
+This sample project is a nice reference architecture to stream data from a
+relational database into an optimized timeseries database. For existing projects
+that are using PostgreSQL, Debezium can be configured to start streaming data to
+QuestDB to take advantage of timeseries queries and partitioning. For databases
+that are also storing raw historical data, some architectural changes are
+needed, but will be an opportunity to improve performance and establish service
+boundaries between a transactional database and an analytical, time-series
+database.
```
Please reword. Difficult to read.

amunra commented on blog/2022-12-28-change-data-capture-with-questdb-and-debezium.mdx:
```diff
@@ -0,0 +1,332 @@
+also stream to other data warehouses for long-term storage. After inspecting the
+data, QuestDB can also be configured to downsample the data for longer term
+storage or
+even[ detach partitions to save space](/blog/2022/11/02/data-lifecycle-questdb/).
```
The space should be moved to before the `[`.

amyshwang replied:
```suggestion
even [detach partitions to save space](/blog/2022/11/02/data-lifecycle-questdb/).
```

amunra reviewed (dismissed):
A pretty useful article, but it still needs some polishing.

amyshwang commented on blog/2022-12-28-change-data-capture-with-questdb-and-debezium.mdx:
```diff
@@ -0,0 +1,332 @@
+
+## Final result
+
+Now all the updates written to the PostgreSQL table will be also reflected in
```
```suggestion
Now all the updates written to the PostgreSQL table will also be reflected in
```

amyshwang commented on blog/2022-12-28-change-data-capture-with-questdb-and-debezium.mdx:
```diff
@@ -0,0 +1,332 @@
+## Final result
+
+Now all the updates written to the PostgreSQL table will be also reflected in
+QuestDB. To validate, navigate to localhost:19000 and select from the stock
```
It's intentional to separate from any other instance the user may have!

amyshwang commented on blog/2022-12-28-change-data-capture-with-questdb-and-debezium.mdx:
```diff
@@ -0,0 +1,332 @@
+
+The
+[code](https://github.com/questdb/kafka-questdb-connector/blob/main/kafka-questdb-connector-samples/stocks/src/main/resources/schema.sql)
+for the Java Spring App can be found under src. It's a simple Spring Boot app
```
```suggestion
for the Java Spring App can be found under `src`. It's a simple Spring Boot app
```

amyshwang commented on blog/2022-12-28-change-data-capture-with-questdb-and-debezium.mdx:
```diff
@@ -0,0 +1,332 @@
+The
+[code](https://github.com/questdb/kafka-questdb-connector/blob/main/kafka-questdb-connector-samples/stocks/src/main/resources/schema.sql)
+for the Java Spring App can be found under src. It's a simple Spring Boot app
+that first creates a new stock table upon startup and inserts symbol , price ,
+and last_update timestamps:
```
```suggestion
that first creates a new stock table upon startup and inserts `symbol`, `price`,
and `last_update` timestamps:
```

amyshwang commented on blog/2022-12-28-change-data-capture-with-questdb-and-debezium.mdx:
````diff
@@ -0,0 +1,332 @@
+    "connector.class": "io.debezium.connector.postgresql.PostgresConnector",
+    "database.dbname": "postgres",
+    "database.server.name": "dbserver1"
+  }
+```
````
```suggestion
  }
 }
```
```

amyshwang commented on blog/2022-12-28-change-data-capture-with-questdb-and-debezium.mdx:
````diff
@@ -0,0 +1,332 @@
+
+The CURL command posted earlier establishes this connection:
+
+```shell
````
```jason

amyshwang replied:
```suggestion
```jason
```

amyshwang commented on blog/2022-12-28-change-data-capture-with-questdb-and-debezium.mdx:
```diff
@@ -0,0 +1,332 @@
+The
+[code](https://github.com/questdb/kafka-questdb-connector/blob/main/kafka-questdb-connector-samples/stocks/src/main/resources/schema.sql)
+for the Java Spring App can be found under src. It's a simple Spring Boot app
+that first creates a new stock table upon startup and inserts symbol , price ,
+and last_update timestamps:
```
```suggestion
that first creates a new stock table upon startup and inserts `symbol`, `price`,
and `last_update` timestamps:
```

amyshwang commented on blog/2022-12-28-change-data-capture-with-questdb-and-debezium.mdx:
````diff
@@ -0,0 +1,332 @@
+
+The CURL command posted earlier establishes this connection:
+
+```shell
````
This is an explanation of the command at L104.

amyshwang commented on blog/2022-12-28-change-data-capture-with-questdb-and-debezium.mdx:
````diff
@@ -0,0 +1,332 @@
+    "include.key": "false",
+    "symbols": "symbol",
+    "timestamp.field.name": "last_update"
+  }
+```
````
```suggestion
  }
 }
```
```

amunra replied:
Mind: Accidentally deleted triple-backtick.

amyshwang commented on blog/2022-12-28-change-data-capture-with-questdb-and-debezium.mdx:
````diff
@@ -0,0 +1,332 @@
+
+The CURL command posted earlier establishes this connection:
+
+```shell
````
```suggestion
```jason
```

amunra replied:
json, not jason

amyshwang commented on blog/2022-12-28-change-data-capture-with-questdb-and-debezium.mdx:
```diff
@@ -0,0 +1,332 @@
+    price float8,
+    last_update timestamp
+);
+INSERT INTO stock (symbol, price, last_update) VALUES ('AAPL', 500.0, now()) ON CONFLICT DO NOTHING;
+INSERT INTO stock (symbol, price, last_update) VALUES ('IBM', 50.0, now()) ON CONFLICT DO NOTHING;
+INSERT INTO stock (symbol, price, last_update) VALUES ('MSFT', 100.0, now()) ON CONFLICT DO NOTHING;
+INSERT INTO stock (symbol, price, last_update) VALUES ('GOOG', 1000.0, now()) ON CONFLICT DO NOTHING;
+INSERT INTO stock (symbol, price, last_update) VALUES ('FB', 200.0, now()) ON CONFLICT DO NOTHING;
+INSERT INTO stock (symbol, price, last_update) VALUES ('AMZN', 1000.0, now()) ON CONFLICT DO NOTHING;
+INSERT INTO stock (symbol, price, last_update) VALUES ('TSLA', 500.0, now()) ON CONFLICT DO NOTHING;
+INSERT INTO stock (symbol, price, last_update) VALUES ('NFLX', 500.0, now()) ON CONFLICT DO NOTHING;
+INSERT INTO stock (symbol, price, last_update) VALUES ('TWTR', 50.0, now()) ON CONFLICT DO NOTHING;
+INSERT INTO stock (symbol, price, last_update) VALUES ('SNAP', 10.0, now()) ON CONFLICT DO NOTHING;
```
I don't have an answer but this is copied form the [code](https://github.com/questdb/kafka-questdb-connector/blob/main/kafka-questdb-connector-samples/stocks/src/main/resources/schema.sql)...

jerrinot replied:
The sample code uses `ON CONFLICT DO NOTHING;` to avoid startup failures when starting the demo multiple times. `Symbol` is a primary key.

amyshwang commented on blog/2022-12-28-change-data-capture-with-questdb-and-debezium.mdx:
````diff
@@ -0,0 +1,340 @@
+[Dockerfile](https://github.com/questdb/kafka-questdb-connector/blob/main/kafka-questdb-connector-samples/stocks/Dockerfile-App)
+and linked to PostgreSQL:
+
+```rust title="Docker-App"
````
```suggestion
``` title="Dockerfile-App"
```

amunra commented on blog/2022-12-28-change-data-capture-with-questdb-and-debezium.mdx:
````diff
@@ -0,0 +1,332 @@
+from some external source. This is packaged into a minimal Dockerfile and linked
+to PostgreSQL:
+
+```sql
````
This is just describing the Dockerfile-App file, so ignore me. Just the title does it.

amunra commented:
The article doesn't flow linearly building up step by step, instead it "does a thing" and then explains it.

This means, IMO that we need better labelling and structure:

- The `curl` commands need individual names and titles so we can refer to them clearly.
- The "Deep dive" section needs a basic outline of the sub portions of the code we're describing for more context.
- In the description of the last `curl` command there needs to be a better explaination of the sample ticking data. There should be a sentence explaning it and -- ideally -- where one would observe it (some log file, perhaps?).


Additionally:
- Why do we have `ON COMMIT DO NOTHING` in the insert SQL? It's confusing and seems unecessary (given there's a `now()` function being called.)

amyshwang commented:
I have addressed the latest [comments](https://github.com/questdb/questdb.io/pull/1244#issuecomment-1367868209) added by @amunra. I added a file structure for the deep dive and more linking for the flow of the content.
However, some outstanding items that I am not sure about - @jerrinot I hope you could add more input:

> In the description of the last curl command there needs to be a better explanation of the sample ticking data. There should be a sentence explaining it and -- ideally -- where one would observe it (some log file, perhaps?).

This is referring to L311 "Example of a JSON message emitted by Debezium". I added more explanation but I don't know where this can be observed 

> Why do we have ON COMMIT DO NOTHING in the insert SQL? It's confusing and seems unecessary (given there's a now() function being called.)

I don't have the answer to this we need to either:
- explain why it's needed
- remove them from both here and the original repo to ensure consistency

jerrinot commented:
indeed, the last JSON is confusing and deserves to be better explained. 

I tried to explain it at the [original code sample](https://github.com/questdb/kafka-questdb-connector/tree/main/kafka-questdb-connector-samples/stocks#kafka-questdb-connector), but the explanation part is missing from this blog. I guess it's because the author deemed it as too detailed. let me see if I can make it concise while improving clarity.

amyshwang commented on blog/2023-01-03-change-data-capture-with-questdb-and-debezium.mdx:
```diff
@@ -0,0 +1,437 @@
+    The updates are not perfectly random, the application uses a very simple algorithm to generate updates which very roughly resembles stock price movements.
+    In a real-life scenario, the application would fetch the price from some external source.
+
+This is packaged into a minimal Dockerfile,
```
What is "This"?

jerrinot replied:
It follows right after this sentence:

![image](https://user-images.githubusercontent.com/158619/210372706-727ca9b4-1821-4903-90b7-7c37cfa7439a.png)

Do you think it's unclear?

amyshwang replied:
Not as clear as it could be...because something is packaged into a dockerfile, and it cannot be the same thing we introduce? Looking at the context, I think we mean No. 2 of the list -- Java code?

amyshwang replied:
I did not express myself well. The line is following a numbered list, so naturally I would ask if "This" means No. 2 - Java code.

amyshwang replied:
```suggestion
The producer is packaged into a minimal Dockerfile,
```

amyshwang commented on blog/2023-01-03-change-data-capture-with-questdb-and-debezium.mdx:
````diff
@@ -0,0 +1,437 @@
+CMD ["java", "-jar", "/stocks.jar"]
+```
+
+### Kafka Connect, Debezium and QuestDB Kafka Connector
````
```suggestion
### Kafka Connect, Debezium, and QuestDB Kafka Connector
```

amyshwang commented on blog/2023-01-03-change-data-capture-with-questdb-and-debezium.mdx:
````diff
@@ -0,0 +1,437 @@
+```
+
+### Kafka Connect, Debezium and QuestDB Kafka Connector
+Before we dive into the Kafka Connect, Debezium and the QuestDB Kafka connector configurations, let's take a look at the relation
````
```suggestion
Before we dive into the Kafka Connect, Debezium, and the QuestDB Kafka connector configurations, let's take a look at their relation
```

amyshwang commented on blog/2023-01-03-change-data-capture-with-questdb-and-debezium.mdx:
```diff
@@ -0,0 +1,437 @@
+
+### Kafka Connect, Debezium and QuestDB Kafka Connector
+Before we dive into the Kafka Connect, Debezium and the QuestDB Kafka connector configurations, let's take a look at the relation
+between Kafka Connect, Debezium and QuestDB Kafka connector.
```
```suggestion
with each other.
```

amyshwang commented on blog/2023-01-03-change-data-capture-with-questdb-and-debezium.mdx:
```diff
@@ -0,0 +1,437 @@
+2. Sink connectors - read data from Kafka and write it to a sink system
+
+Debezium is a Source connector for Kafka Connect that can monitor and capture the row-level changes in the databases.
+What does it mean? Whenever a row is inserted, updated or deleted in a database, Debezium will capture the change and write it as an event to Kafka.
```
```suggestion
What does it mean? Whenever a row is inserted, updated, or deleted in a database, Debezium will capture the change and write it as an event to Kafka.
```

amyshwang commented on blog/2023-01-03-change-data-capture-with-questdb-and-debezium.mdx:
```diff
@@ -0,0 +1,437 @@
+What does it mean? Whenever a row is inserted, updated or deleted in a database, Debezium will capture the change and write it as an event to Kafka.
+
+On a technical level, Debezium is a Kafka Connect connector running inside the Kafka Connect framework.
+This is reflected in the Debezium container image: It packages the Kafka Connect with Debezium connectors pre-installed.
```
Here, "the Debezium contain image" is unclear to me. Is it the image built by [`Dockerfile-Connect`](https://github.com/questdb/kafka-questdb-connector/blob/main/kafka-questdb-connector-samples/stocks/Dockerfile-Connect)? Maybe we could refer to `Dockerfile-Connect`?

amyshwang commented on blog/2023-01-03-change-data-capture-with-questdb-and-debezium.mdx:
```diff
@@ -0,0 +1,437 @@
+This is reflected in the Debezium container image: It packages the Kafka Connect with Debezium connectors pre-installed.
+
+QuestDB Kafka connector is also a Kafka Connect connector. It's a Sink connector that reads data from Kafka and writes it to QuestDB.
+We add the QuestDB Kafka connector to the Debezium container image, and we get a Kafka Connect image that has both Debezium and QuestDB Kafka connector installed!
```
Again, "the Debezium container image" can be confusing; could we say the `Dockerfile-Connect` image?

jerrinot replied:
The problem is that "Dockerfile-Connect" is just a file name. It's not a name of the image. It's a recipe. 
It does exactly what the sentence says: It layers the QuestDB Kafka connector on top of the existing Debezium container image.

amyshwang commented on blog/2023-01-03-change-data-capture-with-questdb-and-debezium.mdx:
```diff
@@ -0,0 +1,437 @@
+The Dockerfile downloads the latest release of the QuestDB Kafka connector, unzip it copies it to the Debezium container image.
+The resulting image has both Debezium and QuestDB Kafka connector installed.
+
+If you are still struggling to grasp all this complexity then this picture should make it clearer:
+
+TBD
```
Is this for the diagram? Currently an overview diagram is in L63: I thought we are updating that one?

jerrinot replied:
yes. as discussed over slack.

amyshwang commented on blog/2023-01-03-change-data-capture-with-questdb-and-debezium.mdx:
```diff
@@ -0,0 +1,437 @@
+So after we send the request, Debezium will start listening to changes in the `stock` table and publish them to the `dbserver1.public.stock` topic.
+
+### QuestDB Kafka Connector
+At this point we have a PostgreSQL table `stock` being populated with random stock prices and a Kafka topic `dbserver1.public.stock` that contains the changes.
```
```suggestion
At this point, we have a PostgreSQL table `stock` being populated with random stock prices and a Kafka topic `dbserver1.public.stock` that contains the changes.
```

amyshwang commented on blog/2023-01-03-change-data-capture-with-questdb-and-debezium.mdx:
````diff
@@ -0,0 +1,437 @@
+}
+```
+Don't get scared if you feel overwhelmed by the sheer size of this message. Most of the fields are metadata, and they are not relevant for this sample.
+See [Debezium documentation](https://debezium.io/documentation/reference/1.9/connectors/postgresql.html#postgresql-events) you if you are interested in details.
````
```suggestion
See [Debezium documentation](https://debezium.io/documentation/reference/1.9/connectors/postgresql.html#postgresql-events), for more details.
```

amyshwang commented on blog/2023-01-03-change-data-capture-with-questdb-and-debezium.mdx:
````diff
@@ -0,0 +1,437 @@
+```
+Don't get scared if you feel overwhelmed by the sheer size of this message. Most of the fields are metadata, and they are not relevant for this sample.
+See [Debezium documentation](https://debezium.io/documentation/reference/1.9/connectors/postgresql.html#postgresql-events) you if you are interested in details.
+The important realization is this: We cannot use the whole JSON message as it is and push it to QuestDB. We do not want all the metadata in QuestDB.
````
```suggestion
The important point is that we cannot push the whole JSON message to QuestDB and we do not want all the metadata in QuestDB.
```

amyshwang reviewed (commented):
Asking some questions and making some minor suggestions!

jerrinot reviewed (approved):
let's publish it!

amunra reviewed (approved):
Nice!
